### PR TITLE
CI: Github actions

### PR DESCRIPTION
**Pull request by tmhuysen**
**Short description**
Perform continuous integration using Github actions

### CODE REVIEW

**Comment by lelemmen** [comment]
🎉

File: .github/workflows/gqcp.yml [hunk 1]
```diff
@@ -0,0 +1,114 @@
+name: GQCP
+
+on: [pull_request]
+
+# These variables are for the slack notification, the default is failure, if all build steps are successful they will be updated to green and "Build successful"
+env:
+  BUILD_COLOR: '#FF0000'
+  BUILD_MESSAGE: 'Build failed'
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]: I'd like to propose the following changes:

```
COLOR_SUCCESSFUL = "#00FF00"  # green
COLOR_FAILED = "#FF000"  # red
MESSAGE_SUCCESSFUL = "Build successful"
MESSAGE_FAILED = "Build failed"

SLACK_COLOR = COLOR_FAILED
SLACK_MESSAGE = SLACK_MESSAGE_FAILED
```

File: .github/workflows/gqcp.yml [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+# Jobs are ran in parallel
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
```

Not sure this is needed

File: .github/workflows/gqcp.yml [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+jobs:
+
+# BUILD
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
```

idem

File: .github/workflows/gqcp.yml [hunk 4]
```diff
@@ -0,0 +1,114 @@
+name: GQCP
+
+on: [pull_request]
+
+# These variables are for the slack notification, the default is failure, if all build steps are successful they will be updated to green and "Build successful"
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
```

Are these variables defined in the same environment as the ones in jobs:build:env?

File: .github/workflows/gqcp.yml [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+  BUILD_COLOR: '#FF0000'
+  BUILD_MESSAGE: 'Build failed'
+
+# Jobs are ran in parallel
+jobs:
+
+# BUILD
+  build:
+    runs-on: ubuntu-latest
+    # Environment variables
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion

    # Environment variables for the build environment
```

File: .github/workflows/gqcp.yml [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    env:
+      CONDA: /usr/share/miniconda
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
      CONDA_HOME: /usr/share/miniconda
```

File: .github/workflows/gqcp.yml [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+      LIBINT_DATA_PATH: $CONDA/share/libint/2.3.1/basis
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
      LIBINT_DATA_PATH: ${CONDA_HOME}/share/libint/2.3.1/basis
```

File: .github/workflows/gqcp.yml [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    - uses: actions/checkout@v1
+
+# Set cache actions: links files at a path to a key to be stored
+#    - name: Cache Conda
+#      id: cache-conda  # cache id
+#      uses: actions/cache@v1.0.3. # Marketplace script
+      # Store all files in $CONDA, link with key "Linux-conda"
+#      with:
+#        path: $CONDA
+#        key: ${{ runner.os }}-conda
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]: If it's commented out, it doesn't belong in the source code

File: .github/workflows/gqcp.yml [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+# BUILD
+  build:
+    runs-on: ubuntu-latest
+    # Environment variables
+    env:
+      CONDA: /usr/share/miniconda
+      LIBINT_DATA_PATH: $CONDA/share/libint/2.3.1/basis
+
+# Various steps to perform the build
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
```

File: .github/workflows/gqcp.yml [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    steps:
+    # Cmake & Python scripts
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion

    # Install CMake and python through GitHub actions
```

File: .github/workflows/gqcp.yml [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+#    - name: Cache Conda
+#      id: cache-conda  # cache id
+#      uses: actions/cache@v1.0.3. # Marketplace script
+      # Store all files in $CONDA, link with key "Linux-conda"
+#      with:
+#        path: $CONDA
+#        key: ${{ runner.os }}-conda
+
+    - name: Install Miniconda
+ #     if: steps.cache-conda.outputs.cache-hit != 'true'  # If cached skip
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
 #     if: steps.cache-conda.outputs.cache-hit != 'true'  # If cached skip
```

idem

File: .github/workflows/gqcp.yml [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+      run: |
+        rm -rf $CONDA;
+        mkdir -p $HOME/download;
+        wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O $HOME/download/miniconda.sh;
+        bash $HOME/download/miniconda.sh -b -p $CONDA;
+
+    - name: Configure Conda
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]: I would merge this into the previous step: "Install and configure Conda"

File: .github/workflows/gqcp.yml [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+      run: |   
+        source $CONDA/bin/activate
+        hash -r
+        conda config --set always_yes yes --set changeps1 no
+        
+    - name: Install Conda Modules
+#      if: steps.cache-conda.outputs.cache-hit != 'true'  # If cached skip
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]: idem

File: .github/workflows/gqcp.yml [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+      run: |   
+        conda install -c conda-forge cmake=3.13 boost-cpp=1.69.0 eigen blas mkl pybind11 benchmark numpy;
+        conda install -c gqcg libint spectra;
+        conda install -c intel mkl-include mkl-static intel-openmp;
+
+    - name: Install Libcint
+#      if: steps.cache-conda.outputs.cache-hit != 'true'  # If cached skip
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]: idem

File: .github/workflows/gqcp.yml [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+        wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O $HOME/download/miniconda.sh;
+        bash $HOME/download/miniconda.sh -b -p $CONDA;
+
+    - name: Configure Conda
+      run: |   
+        source $CONDA/bin/activate
+        hash -r
+        conda config --set always_yes yes --set changeps1 no
+        
+    - name: Install Conda Modules
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
    - name: Install some GQCP dependencies through Conda
```

File: .github/workflows/gqcp.yml [hunk 16]
```diff
@@ -0,0 +1,114 @@
...
+#      if: steps.cache-conda.outputs.cache-hit != 'true'  # If cached skip
+      run: |   
+        conda install -c conda-forge cmake=3.13 boost-cpp=1.69.0 eigen blas mkl pybind11 benchmark numpy;
+        conda install -c gqcg libint spectra;
+        conda install -c intel mkl-include mkl-static intel-openmp;
+
+    - name: Install Libcint
+#      if: steps.cache-conda.outputs.cache-hit != 'true'  # If cached skip
+      # Note that we install Libcint in $CONDA
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
      # Note that we install Libcint in ${CONDA_HOME}
```

File: .github/workflows/gqcp.yml [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+      run: |
+        cd /tmp
+        git clone -b develop https://github.com/GQCG/libcint.git
+        cd libcint
+        mkdir build && cd build
+        $CONDA/bin/cmake .. -DUSE_MKL=TRUE -DCMAKE_INSTALL_PREFIX=$CONDA -DCMAKE_PREFIX_PATH=$CONDA
+        make -j3 && sudo make install
+# Build & test the library
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion

# Build & test the library 
```

match indentation

File: .github/workflows/gqcp.yml [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+        $CONDA/bin/cmake .. -DUSE_MKL=TRUE -DCMAKE_INSTALL_PREFIX=$CONDA -DCMAKE_PREFIX_PATH=$CONDA
+        make -j3 && sudo make install
+# Build & test the library 
+    - name: Build GQCP
+      run: |
+        mkdir build && cd build
+        $CONDA/bin/cmake .. -DBUILD_TESTS=TRUE -DBUILD_BENCHMARKS=TRUE -DBUILD_PYTHON_BINDINGS=TRUE -DCMAKE_INSTALL_PREFIX=~/.local -DCMAKE_PREFIX_PATH=$CONDA
+        make -j10 && sudo make install
+        
+# Run the benchmarks
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]: idem

File: .github/workflows/gqcp.yml [hunk 19]
```diff
@@ -0,0 +1,114 @@
...
+        ./doci_case --benchmark_counters_tabular=true
+        ./doci_matrix --benchmark_counters_tabular=true
+        ./doci_matvec --benchmark_counters_tabular=true
+        ./fci_matrix --benchmark_counters_tabular=true
+        ./fci_matvec --benchmark_counters_tabular=true
+        ./hubbard_diagonalization --benchmark_counters_tabular=true
+        ./hubbard_matrix --benchmark_counters_tabular=true
+        ./hubbard_matvec --benchmark_counters_tabular=true
+
+# Test if installed library links properly to external driver
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
# Test if the installed library links properly to external driver
```
+ match indentation

File: .github/workflows/gqcp.yml [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+      # Note that we install Libcint in $CONDA
+      run: |
+        cd /tmp
+        git clone -b develop https://github.com/GQCG/libcint.git
+        cd libcint
+        mkdir build && cd build
+        $CONDA/bin/cmake .. -DUSE_MKL=TRUE -DCMAKE_INSTALL_PREFIX=$CONDA -DCMAKE_PREFIX_PATH=$CONDA
+        make -j3 && sudo make install
+# Build & test the library 
+    - name: Build GQCP
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
    - name: Build GQCP and run tests
```

File: .github/workflows/gqcp.yml [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+      run: |
+        mkdir build && cd build
+        $CONDA/bin/cmake .. -DBUILD_TESTS=TRUE -DBUILD_BENCHMARKS=TRUE -DBUILD_PYTHON_BINDINGS=TRUE -DCMAKE_INSTALL_PREFIX=~/.local -DCMAKE_PREFIX_PATH=$CONDA
+        make -j10 && sudo make install
+        
+# Run the benchmarks
+    - name: Benchmarks
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
    - name: Run benchmarks
```

File: .github/workflows/gqcp.yml [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+      run: |
+        cd build
+        cd benchmarks
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]: merge this

File: .github/workflows/gqcp.yml [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+        ./doci_matvec --benchmark_counters_tabular=true
+        ./fci_matrix --benchmark_counters_tabular=true
+        ./fci_matvec --benchmark_counters_tabular=true
+        ./hubbard_diagonalization --benchmark_counters_tabular=true
+        ./hubbard_matrix --benchmark_counters_tabular=true
+        ./hubbard_matvec --benchmark_counters_tabular=true
+
+# Test if installed library links properly to external driver
+# This is the final step to verify success, if it does not error it updates the env variables
+    - name: Link
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
    - name: Check external linking
```

File: .github/workflows/gqcp.yml [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+        ./doci_matrix --benchmark_counters_tabular=true
+        ./doci_matvec --benchmark_counters_tabular=true
+        ./fci_matrix --benchmark_counters_tabular=true
+        ./fci_matvec --benchmark_counters_tabular=true
+        ./hubbard_diagonalization --benchmark_counters_tabular=true
+        ./hubbard_matrix --benchmark_counters_tabular=true
+        ./hubbard_matvec --benchmark_counters_tabular=true
+
+# Test if installed library links properly to external driver
+# This is the final step to verify success, if it does not error it updates the env variables
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
# This is the final step to verify a successful build: if it does not error, it updates the environment variables to be used in the Slack notification
```

File: .github/workflows/gqcp.yml [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+        cd /tmp
+        git clone https://github.com/GQCG/gqcp-link.git
+        cd gqcp-link
+        mkdir build && cd build
+        $CONDA/bin/cmake .. -DCMAKE_PREFIX_PATH=$CONDA -DGQCP_INSTALL_PREFIX=~/.local -Dgqcp_DIR=/home/runner/.local/cmake
+        make -j3 && ./test_driver
+        echo "::set-env name=BUILD_MESSAGE::Build successful" 
+        echo "::set-env name=BUILD_COLOR::#00FF00"
+        
+# Slack Notifications
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
```

File: .github/workflows/gqcp.yml [hunk 26]
```diff
@@ -0,0 +1,114 @@
...
+    - name: slackNotification
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
    - name: Slack notification
```

File: .github/workflows/gqcp.yml [hunk 27]
```diff
@@ -0,0 +1,114 @@
...
+      if: always()  # This step will always run, even if a previous step fails
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
      if: always()  # this step will always run, even if a previous step fails
```

File: .github/workflows/gqcp.yml [hunk 28]
```diff
@@ -0,0 +1,114 @@
...
+      uses: rtCamp/action-slack-notify@master
+      env:
+        SLACK_WEBHOOK: ${{ secrets.SLACK_WEBHOOK }}
+        SLACK_COLOR: ${{ env.BUILD_COLOR}}
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
        SLACK_COLOR: ${{ env.SLACK_COLOR }}
```

File: .github/workflows/gqcp.yml [hunk 29]
```diff
@@ -0,0 +1,114 @@
...
+        SLACK_ICON: https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png
+        SLACK_MESSAGE: ${{ env.BUILD_MESSAGE}}
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
        SLACK_MESSAGE: ${{ env.SLACK_MESSAGE }}
```

File: .github/workflows/gqcp.yml [hunk 30]
```diff
@@ -0,0 +1,114 @@
...
+        SLACK_TITLE: ${{ env.GITHUB_REF }}
+        SLACK_USERNAME: GitHub
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
        SLACK_USERNAME: GitHub Actions
```

File: .github/workflows/gqcp.yml (hunk 1 above)
Comment (tmhuysen) [reply .github/workflows/gqcp.yml]: The slack environment only exists after "using" the action. 
`uses: rtCamp/action-slack-notify@master` However I want to update them these variables outside of the slack step, so I pre-initialize variables containing the information I want, update them (potentially) and then initialize the slack variables with those values later on.

File: .github/workflows/gqcp.yml (hunk 4 above)
Comment (tmhuysen) [reply .github/workflows/gqcp.yml]: I actually didn't test if it would work that way.

File: .github/workflows/gqcp.yml (hunk 12 above)
Comment (tmhuysen) [reply .github/workflows/gqcp.yml]: This is separated for caching

File: .github/workflows/gqcp.yml (hunk 1 above)
Comment (tmhuysen) [reply .github/workflows/gqcp.yml]: Okay, I understand what you mean here now. I don't know if the variables 'jobs.build.Slack%20Notifications.env.SLACK_MESSAGE` and `env.SLACK_MESSAGE` can both exist, we will see.

File: .github/workflows/gqcp.yml [hunk 31]
```diff
@@ -0,0 +1,104 @@
...
+
+on: [pull_request]
+
+# These variables are for the slack notification, the default is failure, if all build steps are successful they will be updated to green and "Build successful"
+env:
+  COLOR_SUCCESSFUL: "#00FF00"  # green
+  COLOR_FAILED: "#FF000"  # red
+  MESSAGE_SUCCESSFUL: Build successful
+  MESSAGE_FAILED: Build failed
+# Defaults are failure
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion

# Defaults are failure
```

plus add indentation to match `SLACK_COLOR`

File: .github/workflows/gqcp.yml [hunk 32]
```diff
@@ -0,0 +1,104 @@
...
+  SLACK_COLOR: COLOR_FAILED  
+  SLACK_MESSAGE: MESSAGE_FAILED
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    # Environment variables for the build environment (build.env)
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion

    # Environment variables for the build environment (build.env)
```

File: .github/workflows/gqcp.yml [hunk 33]
```diff
@@ -0,0 +1,104 @@
...
+        wget https://repo.continuum.io/miniconda/Miniconda3-latest-Linux-x86_64.sh -O $HOME/download/miniconda.sh;
+        bash $HOME/download/miniconda.sh -b -p ${CONDA_HOME};
+
+    - name: Configure Conda
+      run: |   
+        source ${CONDA_HOME}/bin/activate
+        hash -r
+        conda config --set always_yes yes --set changeps1 no
+        
+
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
```

File: .github/workflows/gqcp.yml [hunk 34]
```diff
@@ -0,0 +1,104 @@
...
+      # Note that we install Libcint in ${CONDA_HOME}
+      run: |
+        cd /tmp
+        git clone -b develop https://github.com/GQCG/libcint.git
+        cd libcint
+        mkdir build && cd build
+        ${CONDA_HOME}/bin/cmake .. -DUSE_MKL=TRUE -DCMAKE_INSTALL_PREFIX=${CONDA_HOME} -DCMAKE_PREFIX_PATH=${CONDA_HOME}
+        make -j3 && sudo make install
+        
+# Build & test the library
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]: Match indentation of the comment to `- name`

File: .github/workflows/gqcp.yml [hunk 35]
```diff
@@ -0,0 +1,104 @@
...
+        make -j3 && sudo make install
+        
+# Build & test the library 
+    - name: Build GQCP and run tests
+      run: |
+        mkdir build && cd build
+        ${CONDA_HOME}/bin/cmake .. -DBUILD_TESTS=TRUE -DBUILD_BENCHMARKS=TRUE -DBUILD_PYTHON_BINDINGS=TRUE -DCMAKE_INSTALL_PREFIX=~/.local -DCMAKE_PREFIX_PATH=${CONDA_HOME}
+        make -j10 && sudo make install
+        
+# Run the benchmarks
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]: Match indentation of the comment to `- name`

File: .github/workflows/gqcp.yml [hunk 36]
```diff
@@ -0,0 +1,104 @@
...
+        ./doci_matrix --benchmark_counters_tabular=true
+        ./doci_matvec --benchmark_counters_tabular=true
+        ./fci_matrix --benchmark_counters_tabular=true
+        ./fci_matvec --benchmark_counters_tabular=true
+        ./hubbard_diagonalization --benchmark_counters_tabular=true
+        ./hubbard_matrix --benchmark_counters_tabular=true
+        ./hubbard_matvec --benchmark_counters_tabular=true
+
+# Test if the installed library links properly to external driver
+# This is the final step to verify a successful build: if it does not error it updates the env variables
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]: Match indentation of the comments to `- name`

File: .github/workflows/gqcp.yml [hunk 37]
```diff
@@ -0,0 +1,104 @@
...
+        make -j3 && ./test_driver
+        echo "::set-env name=SLACK_MESSAGE::$MESSAGE_SUCCESSFUL" 
+        echo "::set-env name=SLACK_COLOR::$COLOR_SUCCESSFUL"
+        
+    - name: Slack notification
+      if: always()  # this step will always run, even if a previous step fails
+      uses: rtCamp/action-slack-notify@master
+      env:
+        SLACK_WEBHOOK: ${{ secrets.SLACK_WEBHOOK }}
+        SLACK_COLOR: ${{ env.SLACK_COLOR}}
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
        SLACK_COLOR: ${{ env.SLACK_COLOR }}
```

File: .github/workflows/gqcp.yml [hunk 38]
```diff
@@ -0,0 +1,104 @@
...
+        SLACK_ICON: https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png
+        SLACK_MESSAGE: ${{ env.SLACK_MESSAGE}}
```
Comment (lelemmen) [inline .github/workflows/gqcp.yml]:
```suggestion
        SLACK_MESSAGE: ${{ env.SLACK_MESSAGE }}
```

**Pull request merged**